### PR TITLE
chore(asserts_custom_model_rules): migrate to Terraform Framework SDK

### PR DESCRIPTION
**Pull request by leonorfmartins**
Migrates `grafana_asserts_custom_model_rules` from the legacy Terraform Plugin SDK v2 to the Terraform Plugin
  Framework.

  - Replaces `map[string]interface{}` data handling with typed Go structs (`customModelRulesModel`,
  `rulesModel`, `entityModel`, `definedByModel`).
  - Implements `resource.Resource`, `resource.ResourceWithConfigure`, and `resource.ResourceWithImportState`
  interfaces.
  - Replaces the SDK-coupled `withRetryRead` helper with a Framework-compatible `readModelWithRetry` method
  using exponential backoff with jitter, to handle eventual consistency after write operations.
  - Preserves all existing behavioral semantics, including the `disabled`-only-when-true API contract.
  - Adds comprehensive unit tests covering API↔model conversion, round-trips, disabled field edge cases, and
  error paths.

Fixes https://github.com/grafana/deployment_tools/issues/547404

### CODE REVIEW

**Comment by leonorfmartins** [comment]
@codex review

**Comment by leonorfmartins** [comment]
@codex review

**Comment by leonorfmartins** [comment]
@codex review

**Comment by leonorfmartins** [comment]
@codex review

File: docs/resources/asserts_custom_model_rules.md [hunk 1]
```diff
@@ -65,21 +68,21 @@ resource "grafana_asserts_custom_model_rules" "test" {
 <a id="nestedblock--rules"></a>
 ### Nested Schema for `rules`
 
-Required:
+Optional:
```
Comment (FCalero) [inline docs/resources/asserts_custom_model_rules.md:71]: 🤔 Why were `entity` and `defined_by` moved from required to optional ?

**Review by FCalero — APPROVED** [review]

File: docs/resources/asserts_custom_model_rules.md (hunk 1 above)
Comment (leonorfmartins) [reply docs/resources/asserts_custom_model_rules.md:71]: from what I understand, they seemed optional already in [here](https://github.com/grafana/terraform-provider-grafana/pull/2720/changes#diff-a6d20cd639b183bdcd733e3d81e12e7e5ca34f7e7903ae275d784f7a9bf5d380L98), but I can move them to required if these fields are indeed required :)
